{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter\n",
    "import random\n",
    "import os\n",
    "import subprocess\n",
    "import sys\n",
    "import csv\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Fasta class with similar methods to Biopython, will hold a sequence record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "class fasta:\n",
    "    \n",
    "    GC_content = 0\n",
    "    c = {}\n",
    "    transcription = ''\n",
    "    translation = ''\n",
    "    translation_nospaces = ''\n",
    "    is_dna = True\n",
    "    molecular_weight = float(0)\n",
    "    coronavirus = False\n",
    "    \n",
    "    def __init__(self, name, sequence):\n",
    "        self.name = name\n",
    "        self.sequence = sequence\n",
    "        self.c = Counter(sequence)\n",
    "        self.checkDNA_orProtein()\n",
    "        self.checkCoronavirus()\n",
    "        \n",
    "        if self.is_dna == True:\n",
    "            self.transcription = self.transcribe()\n",
    "            self.translation = self.translate()\n",
    "        else:\n",
    "            self.translation = self.sequence\n",
    "            self.translation_nospaces = self.sequence\n",
    "        \n",
    "    \n",
    "    def checkCoronavirus(self):\n",
    "        if 'coronavirus' in self.name:\n",
    "            self.coronavirus = True\n",
    "        else:\n",
    "            self.coronavirus = False\n",
    "        \n",
    "        \n",
    "    def print_translation(self, sequence):\n",
    "        if len(sequence) > 50:\n",
    "            return sequence[0:50] + \"....\"\n",
    "        else:\n",
    "            return sequence\n",
    "    \n",
    "    def getLength(sequence):\n",
    "        return len(sequence)\n",
    "    \n",
    "    def checkDNA_orProtein(self):\n",
    "        DNA_set = {'A', 'C', 'G', 'T'}\n",
    "        sequence_set = set(list(self.sequence))\n",
    "        difference = sequence_set - DNA_set\n",
    "        \n",
    "        if len(difference) != 0:\n",
    "            self.is_dna = False\n",
    "        \n",
    "    def getAccession(self):\n",
    "        first_index = self.name.index('|')\n",
    "        return self.name[0:first_index]\n",
    "        \n",
    "        \n",
    "    def toString(self):\n",
    "        return self.name + \" \" + self.sequence\n",
    "    \n",
    "    def get_GC_content(self):\n",
    "        G_content = self.c.get('G')\n",
    "        C_content = self.c.get('C')\n",
    "        GC_content = (G_content + C_content)/len(self.sequence)\n",
    "        return \"{:.6%}\".format(GC_content)\n",
    "        \n",
    "    def transcribe(self):\n",
    "        dna_string = self.sequence\n",
    "        dna_string = dna_string.replace('T', 'U')\n",
    "        return dna_string\n",
    "    \n",
    "    def printDNACounter(self):\n",
    "        return 'A = ' + str(self.c.get('A')), 'C = ' + str(self.c.get('C')), 'G = ' + str(self.c.get('G')), 'T = ' + str(self.c.get('T'))\n",
    "    \n",
    "    def getName(self):\n",
    "        first_index = self.name.index('[')\n",
    "        second_index = self.name.index(']')\n",
    "        return self.name[first_index+1:second_index]\n",
    "        \n",
    "    \n",
    "    def translate(self):\n",
    "        rna = self.transcription\n",
    "        codon_table = {\"UUU\":\"F\", \"UUC\":\"F\", \"UUA\":\"L\", \"UUG\":\"L\",\n",
    "            \"UCU\":\"S\", \"UCC\":\"S\", \"UCA\":\"S\", \"UCG\":\"S\",\n",
    "            \"UAU\":\"Y\", \"UAC\":\"Y\", \"UAA\":\"Stop\", \"UAG\":\"Stop\",\n",
    "            \"UGU\":\"C\", \"UGC\":\"C\", \"UGA\":\"Stop\", \"UGG\":\"W\",\n",
    "            \"CUU\":\"L\", \"CUC\":\"L\", \"CUA\":\"L\", \"CUG\":\"L\",\n",
    "            \"CCU\":\"P\", \"CCC\":\"P\", \"CCA\":\"P\", \"CCG\":\"P\",\n",
    "            \"CAU\":\"H\", \"CAC\":\"H\", \"CAA\":\"Q\", \"CAG\":\"Q\",\n",
    "            \"CGU\":\"R\", \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\",\n",
    "            \"AUU\":\"I\", \"AUC\":\"I\", \"AUA\":\"I\", \"AUG\":\"M\",\n",
    "            \"ACU\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\",\n",
    "            \"AAU\":\"N\", \"AAC\":\"N\", \"AAA\":\"K\", \"AAG\":\"K\",\n",
    "            \"AGU\":\"S\", \"AGC\":\"S\", \"AGA\":\"R\", \"AGG\":\"R\",\n",
    "            \"GUU\":\"V\", \"GUC\":\"V\", \"GUA\":\"V\", \"GUG\":\"V\",\n",
    "            \"GCU\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\", \"GCG\":\"A\",\n",
    "            \"GAU\":\"D\", \"GAC\":\"D\", \"GAA\":\"E\", \"GAG\":\"E\",\n",
    "            \"GGU\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\",}\n",
    "            \n",
    "        amino_acid_chain = []\n",
    "        for i in range(0,len(rna),3):\n",
    "            codon = rna[i:i+3]\n",
    "            amino_acid = codon_table.get(codon)\n",
    "            amino_acid_chain.append(amino_acid)\n",
    "        \n",
    "        amino_acid_chain_string = '-'.join(str(x) for x in amino_acid_chain)\n",
    "        self.translation_nospaces = amino_acid_chain_string.split('-')\n",
    "        return amino_acid_chain_string\n",
    "    \n",
    "    def calculateMolecularWeight(self):\n",
    "        aa = {'A' : 71.03711,'R' : 156.10111,'N' : 114.04293,'D' : 115.02694,'C' : 103.00919,'E' : 129.04259,\n",
    "              'Q' : 128.05858,'G' : 57.02146,'H' : 137.05891,'I' : 113.08406,'L' : 113.08406,'K' : 128.09496,\n",
    "              'M' : 131.04049,'F' : 147.06841,'P' : 97.05276,'S' : 87.03203,'T' : 101.04768,'W' : 186.07931,\n",
    "              'Y' : 163.06333,'V' : 99.06841, 'X' : 0.00000  }\n",
    "        protein_sequence = self.translation_nospaces\n",
    "        molecular_weight = float(0)\n",
    "        \n",
    "        for amino_acid in protein_sequence:\n",
    "            aa_weight = aa.get(amino_acid) \n",
    "            if aa_weight != None:\n",
    "                molecular_weight += aa_weight\n",
    "            else:\n",
    "                continue\n",
    "            \n",
    "        self.molecular_weight = molecular_weight\n",
    "        return molecular_weight\n",
    "        \n",
    "    \n",
    "    def printDNA_analysis(self):\n",
    "        print('\\n Name:  ', self.getName())\n",
    "        print('\\n \\t DNA Analysis: ')\n",
    "        print('--------------------------')\n",
    "        print('\\t DNA sequence: ', self.sequence)\n",
    "        print('\\t', 'Transcription: ',self.transcription)\n",
    "        print('\\t', 'Number of each base in DNA: ', self.printDNACounter())\n",
    "        print('\\t GC content: ', self.get_GC_content())\n",
    "    \n",
    "            \n",
    "    def printProtein_analysis(self):\n",
    "        print('\\n \\t Protein Analysis: ')\n",
    "        print('--------------------------')\n",
    "        print('\\n \\t Amino Acid Chain: ', self.print_translation(self.translation))\n",
    "        print('\\t Number of Amino acids: ', len(self.translation_nospaces))\n",
    "        print('\\t Molecular Weight: ', self.calculateMolecularWeight(), 'Da')\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Some Methods you can run on a list of sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "def profile_consensus_matrix(fasta_records):\n",
    "    equal_length = len(fasta_records[0])\n",
    "    count = 0\n",
    "    profile_matrix = []\n",
    "    \n",
    "    for i in range(4):   \n",
    "        row = []\n",
    "        for column in zip(*fasta_records):\n",
    "            if i == 0:\n",
    "                row.append(column.count('A'))\n",
    "                continue\n",
    "            elif i == 1:\n",
    "                row.append(column.count('C'))\n",
    "                continue\n",
    "            elif i == 2:\n",
    "                row.append(column.count('G'))\n",
    "                continue\n",
    "            else:\n",
    "                row.append(column.count('T'))\n",
    "                continue\n",
    "            \n",
    "        profile_matrix.append(row)\n",
    "    \n",
    "    print(\"Profile Matrix\\n\")\n",
    "    for index,value in enumerate(profile_matrix):\n",
    "        if index == 0:\n",
    "            print(\"A: \", end = \" \")\n",
    "        if index == 1:\n",
    "            print(\"C: \", end = \" \")\n",
    "        if index == 2:\n",
    "            print(\"G: \", end = \" \")\n",
    "        if index == 3:\n",
    "            print(\"T: \", end = \" \")\n",
    "        for c in value:\n",
    "            print(c, end = \" \")\n",
    "            \n",
    "        print(\"\\n\")\n",
    "        \n",
    "    consensus_matrix = []\n",
    "    for column in zip(*profile_matrix):\n",
    "        max_num = max(column)\n",
    "        index_of_max = column.index(max_num)\n",
    "    \n",
    "        if index_of_max == 0:\n",
    "            consensus_matrix.append('A')\n",
    "        if index_of_max == 1:\n",
    "            consensus_matrix.append('C')\n",
    "        if index_of_max == 2:\n",
    "            consensus_matrix.append('G')\n",
    "        if index_of_max == 3:\n",
    "            consensus_matrix.append('T')\n",
    "            \n",
    "    print('\\nConsensus Matrix \\n')\n",
    "    for x in consensus_matrix:\n",
    "        print(x, end = \" \")\n",
    "        \n",
    "        \n",
    "def findingSharedMotif(fasta_records):\n",
    "    fasta_records = sorted(fasta_records, key = len)\n",
    "    shortest_seq = fasta_records[0]\n",
    "    \n",
    "    compare_seq = fasta_records[1:]\n",
    "    motif = ''\n",
    "    for x in range(len(shortest_seq)):\n",
    "        for y in range(x, len(shortest_seq)):\n",
    "            current_seq = shortest_seq[x:y+1]\n",
    "            found = False\n",
    "            for seq in compare_seq:\n",
    "                if current_seq in seq:\n",
    "                    found = True\n",
    "                else:\n",
    "                    found = False\n",
    "                    break   \n",
    "            if found == True and len(current_seq) > len(motif):\n",
    "                motif = current_seq\n",
    "                \n",
    "    return motif\n",
    "\n",
    "       "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Printing the analysis and then some file IO."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "f = open('basic_ds.fasta', 'r')\n",
    "f2 = open('related_ds.fasta', 'r')\n",
    "\n",
    "def file_into_fasta_records(f):\n",
    "    fasta_records = []\n",
    "    for seq_record in SeqIO.parse(f, \"fasta\"):\n",
    "        name = seq_record.description\n",
    "        sequence = str(seq_record.seq)\n",
    "        fasta_records.append(fasta(name, sequence))\n",
    "        \n",
    "    print('\\nNumber of records: ', len(fasta_records))\n",
    "    return fasta_records\n",
    "\n",
    "def fasta_records_analysis(fasta_records):\n",
    "    sequence_record = []\n",
    "    for x in fasta_records:\n",
    "        sequence_record.append(x.sequence)\n",
    "    \n",
    "    print(\"Shared Motif is: \", findingSharedMotif(sequence_record))\n",
    "    if x.is_dna == True:\n",
    "        profile_consensus_matrix(fasta_records)\n",
    "    for x in fasta_records:\n",
    "        if x.is_dna == True:\n",
    "            x.printDNA_analysis()\n",
    "            x.printProtein_analysis()\n",
    "        else:\n",
    "            print('\\n Name:  ', x.getName())\n",
    "            x.printProtein_analysis()\n",
    "            \n",
    "def write_fasta(fasta_records, length, file, length_of_first):\n",
    "    if length_of_first == True:\n",
    "        for i in range(0,length):\n",
    "            file.write('\\n\\n>' + fasta_records[i].name)\n",
    "            file.write('\\n' + fasta_records[i].sequence)\n",
    "    else:\n",
    "        for i in range(length, len(fasta_records)):\n",
    "            file.write('\\n\\n>' + fasta_records[i].name)\n",
    "            file.write('\\n' + fasta_records[i].sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Number of records:  26\n",
      "\n",
      "Basic Dataset \n",
      "\n",
      "\n",
      "Shared Motif is:  R\n",
      "\n",
      " Name:   Acholeplasma virus L2\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  DTIHGPDIVHKQTSSIVTINDIILLYESDGRFVSVAEDGYSGNGARVGTY....\n",
      "\t Number of Amino acids:  212\n",
      "\t Molecular Weight:  23228.095139999983 Da\n",
      "\n",
      " Name:   Swine enteric coronavirus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MTFPRALTVIDDNGVVISIIFWFLLIIILILLSIALLNIIKLCMVCCNLG....\n",
      "\t Number of Amino acids:  82\n",
      "\t Molecular Weight:  9266.163529999996 Da\n",
      "\n",
      " Name:   Human papillomavirus type 10\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MSMGAQEPRNILLLCRNCGIPLEDLRLCCIFCTKQLTAAELAAFALRELY....\n",
      "\t Number of Amino acids:  148\n",
      "\t Molecular Weight:  17533.822229999994 Da\n",
      "\n",
      " Name:   Bat Hp-betacoronavirus/Zhejiang2013\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MYSFVSQETGTVIVNAVFILVGFVALLIVALAILTCLRLCAYCCNILDQG....\n",
      "\t Number of Amino acids:  79\n",
      "\t Molecular Weight:  8795.687629999995 Da\n",
      "\n",
      " Name:   Lelystad virus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MGGLDDFCNDPIAAQKLVLAFSITYTPIMIYALKVSRGRLLGLLHILIFL....\n",
      "\t Number of Amino acids:  173\n",
      "\t Molecular Weight:  18895.14139999998 Da\n",
      "\n",
      " Name:   Human coronavirus OC43\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MFMADAYLADTVWYVGQIIFIVAICLLVTIVVVAFLATFKLCIQLCGMCN....\n",
      "\t Number of Amino acids:  84\n",
      "\t Molecular Weight:  9513.868349999997 Da\n",
      "\n",
      " Name:   Bubaline alphaherpesvirus 1\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MLGPESAALLRAPESPLGGAEAEEGAAGAGPGDGGSDGEGEEDELLRCVA....\n",
      "\t Number of Amino acids:  233\n",
      "\t Molecular Weight:  23434.996550000003 Da\n",
      "\n",
      " Name:   Common moorhen coronavirus HKU21\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MVDNWDITIPGDYVIAALVVICVAVFLLFINTCLACIKLVYKCYKGATVL....\n",
      "\t Number of Amino acids:  82\n",
      "\t Molecular Weight:  9180.782620000002 Da\n",
      "\n",
      " Name:   Wobbly possum disease virus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MRYMTPTHLPPVTMSSLTVTSCMKCWGMRRPSWSHSGIIAHTLDT\n",
      "\t Number of Amino acids:  45\n",
      "\t Molecular Weight:  5085.4188699999995 Da\n",
      "\n",
      " Name:   Lymantria dispar multiple nucleopolyhedrovirus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MSFFSSLRRVNKVYPNPNQFLAVDNLRVLNSTPNGFQNVFSAPSTRQIGT....\n",
      "\t Number of Amino acids:  356\n",
      "\t Molecular Weight:  38594.623820000044 Da\n",
      "\n",
      " Name:   Macropodid alphaherpesvirus 1\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MASNSADYLGEDPPSDGGLTIEEHVLLSSYGPPDSWFNSATCESVGLHKQ....\n",
      "\t Number of Amino acids:  215\n",
      "\t Molecular Weight:  23742.46184999998 Da\n",
      "\n",
      " Name:   Murine roseolovirus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MDRIEILAAIYETFYLVCVILLALQCFILMFCNPNCRVKMLGIGMYVSYF....\n",
      "\t Number of Amino acids:  304\n",
      "\t Molecular Weight:  35097.97433999999 Da\n",
      "\n",
      " Name:   Rabbit coronavirus HKU14\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MFMADAYLADTVWYVGQIIFIVAICLLVIIVVVALLATFKLCIQLCGMCN....\n",
      "\t Number of Amino acids:  89\n",
      "\t Molecular Weight:  10026.294469999997 Da\n",
      "\n",
      " Name:   Wencheng Sm shrew coronavirus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MLPLIYNDNGLVVSVFLWFLILIVFLLFLIVFIKCCKCVFLTHRALNATV....\n",
      "\t Number of Amino acids:  75\n",
      "\t Molecular Weight:  8941.810559999996 Da\n",
      "\n",
      " Name:   Equine arteritis virus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MGLVWSLISNSIQTIIADFAISVIDAALFFLMLLALAIVTVFLFWVIVAI....\n",
      "\t Number of Amino acids:  67\n",
      "\t Molecular Weight:  7348.114139999999 Da\n",
      "\n",
      " Name:   Small ruminant lentivirus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  TTGGLLVIICRNNKTITRIMGTNTNWTTMWGIYKKCSECQNATLNRTGTG....\n",
      "\t Number of Amino acids:  83\n",
      "\t Molecular Weight:  9090.442549999994 Da\n",
      "\n",
      " Name:   Dengue virus 3\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  FFNMLKRARNRVSTGSQLAKRFSKGLLNGQGPMKLVMAFIAFLRFLAIPP....\n",
      "\t Number of Amino acids:  97\n",
      "\t Molecular Weight:  10851.144299999993 Da\n",
      "\n",
      " Name:   Zika virus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  LVMILLIAPAYSIRCIGVSNRDFVEGMSGGTWVDVVLEHGGCVTVMAQDK....\n",
      "\t Number of Amino acids:  109\n",
      "\t Molecular Weight:  11879.721559999996 Da\n",
      "\n",
      " Name:   Lumpy skin disease virus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MADIPLYVIPIVGREISDVVPELKSDNDIFYKKVDTVKDFKNSDVNFFFK....\n",
      "\t Number of Amino acids:  87\n",
      "\t Molecular Weight:  10028.134289999998 Da\n",
      "\n",
      " Name:   Chikungunya virus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  FDNKIVVYKGDVYNMDYPPFGAGRPGQFGDIQSRTPESKDVYANTQLVLQ....\n",
      "\t Number of Amino acids:  130\n",
      "\t Molecular Weight:  14129.893749999987 Da\n",
      "\n",
      " Name:   Human immunodeficiency virus 1\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  ENHTTNIRTIIVQLANPVRINCTRSNNNTRKSVRIGPGQTFYATGDIIGD....\n",
      "\t Number of Amino acids:  93\n",
      "\t Molecular Weight:  10167.159639999994 Da\n",
      "\n",
      " Name:   Rubella virus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  FHTETRTVWQLSVAGVSCNVTTEHPFCNTPHGQLEVQVPPDPGDLVEYIM....\n",
      "\t Number of Amino acids:  74\n",
      "\t Molecular Weight:  8172.752899999997 Da\n",
      "\n",
      " Name:   West Nile virus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  AIFVHGPTTVESHGNYSTQIGATQAGRFSITPSAPSYTLKLGEYGEVTVD....\n",
      "\t Number of Amino acids:  106\n",
      "\t Molecular Weight:  11730.607349999987 Da\n",
      "\n",
      " Name:   Hepatitis B virus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  SSISSRTGDPAPTMENITSGFLGPLLVLQAGFFLLTRILTIPQSLDSWWT....\n",
      "\t Number of Amino acids:  104\n",
      "\t Molecular Weight:  11409.810629999993 Da\n",
      "\n",
      " Name:   Hepacivirus C\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  SMQGNWAKVAIIMVMFSGVDATTYATGGNAARNTYGIASLFNQGPKQNLQ....\n",
      "\t Number of Amino acids:  132\n",
      "\t Molecular Weight:  14449.99599 Da\n",
      "\n",
      " Name:   Kunjin virus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  RGWGNGCGLFGKGSIDTCAKFACSTKATGRTILKENIKYEVAIFVHGPTT....\n",
      "\t Number of Amino acids:  134\n",
      "\t Molecular Weight:  14457.043309999986 Da\n",
      "\n",
      "Number of records:  25\n",
      "\n",
      "Related Dataset \n",
      "\n",
      "\n",
      "Shared Motif is:  M\n",
      "\n",
      " Name:   Bat coronavirus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MYEFVSTETSVLITNILLLVIASLLFVIVGCALLLCIQFLIGACGCLANI....\n",
      "\t Number of Amino acids:  75\n",
      "\t Molecular Weight:  8300.51828 Da\n",
      "\n",
      " Name:   Shrew coronavirus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MSFHPFINDDGTIINVILLLVALTVLLIISVAILSTINLCCKALSCCNKI....\n",
      "\t Number of Amino acids:  78\n",
      "\t Molecular Weight:  8464.704459999999 Da\n",
      "\n",
      " Name:   Canada goose coronavirus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MTDILTKSVQENGSFLTAVYVFTAFVALYLLGKALYAFTQAADACFAFWY....\n",
      "\t Number of Amino acids:  100\n",
      "\t Molecular Weight:  11080.763979999994 Da\n",
      "\n",
      " Name:   Severe acute respiratory syndrome coronavirus 2\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MYSFVSEETGTLIVNSVLLFLAFVVFLLVTLAILTALRLCAYCCNIVNVS....\n",
      "\t Number of Amino acids:  75\n",
      "\t Molecular Weight:  8341.524179999999 Da\n",
      "\n",
      " Name:   Human coronavirus OC43\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MFMADAYLADTVWYVGQIIFIVAICLLVTIVVVAFLATFKLCIQLCGMCN....\n",
      "\t Number of Amino acids:  84\n",
      "\t Molecular Weight:  9513.868349999997 Da\n",
      "\n",
      " Name:   Porcine coronavirus HKU15\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MVVDDWAVTIPGQYIIAILVVICIGVALLFINTCLACVKLFYKCYLGAAY....\n",
      "\t Number of Amino acids:  83\n",
      "\t Molecular Weight:  9425.9983 Da\n",
      "\n",
      " Name:   Lucheng Rn rat coronavirus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MLPSFLRVFNDEGLVLSVLFWFLFIIVLLLFSIAMLKTIQLVGVCCSLTN....\n",
      "\t Number of Amino acids:  78\n",
      "\t Molecular Weight:  9018.929659999998 Da\n",
      "\n",
      " Name:   Ferret coronavirus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MKFPTLLTVIDDNGVVVNSIFWLLLIIVIILFSIALLNIIRLCQTCCRLT....\n",
      "\t Number of Amino acids:  82\n",
      "\t Molecular Weight:  9221.983939999996 Da\n",
      "\n",
      " Name:   Camel alphacoronavirus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MFLKLVDDHALVVNVILWCVVLIVLLLVCITIIKLIKLCFTCHMFCNKTI....\n",
      "\t Number of Amino acids:  77\n",
      "\t Molecular Weight:  9025.864289999998 Da\n",
      "\n",
      " Name:   Bat Hp-betacoronavirus/Zhejiang2013\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MYSFVSQETGTVIVNAVFILVGFVALLIVALAILTCLRLCAYCCNILDQG....\n",
      "\t Number of Amino acids:  79\n",
      "\t Molecular Weight:  8795.687629999995 Da\n",
      "\n",
      " Name:   Middle East respiratory syndrome-related coronavirus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MLPFVQERIGLFIVNFFIFTVVCAITLLVCMAFLTATRLCVQCMTGFNTL....\n",
      "\t Number of Amino acids:  82\n",
      "\t Molecular Weight:  9329.842429999997 Da\n",
      "\n",
      " Name:   Mink coronavirus strain WD1127\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MRFPNVLTVIDDNGIVINSIFWLLLIIVIILFSIALLNVVKLCMVCCRLS....\n",
      "\t Number of Amino acids:  82\n",
      "\t Molecular Weight:  9141.994099999998 Da\n",
      "\n",
      " Name:   Rousettus bat coronavirus HKU10\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MMFTLVNDNGMIVSAILWLVVLLFVLLIAVTVIKLIQLCFTCHKLMSNTI....\n",
      "\t Number of Amino acids:  75\n",
      "\t Molecular Weight:  8551.502149999998 Da\n",
      "\n",
      " Name:   Rabbit coronavirus HKU14\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MFMADAYLADTVWYVGQIIFIVAICLLVIIVVVALLATFKLCIQLCGMCN....\n",
      "\t Number of Amino acids:  89\n",
      "\t Molecular Weight:  10026.294469999997 Da\n",
      "\n",
      " Name:   Sparrow coronavirus HKU17\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MVADDWAVTIPGHYIIAILIVVCIGVALLFINTCLACVKLTYKCYLGARY....\n",
      "\t Number of Amino acids:  82\n",
      "\t Molecular Weight:  9318.89966 Da\n",
      "\n",
      " Name:   Magpie-robin coronavirus HKU18\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MVVEDWSITIPGQYVIATLAVLCVCVCLLFINTCLACAKLSYKCFLGAKY....\n",
      "\t Number of Amino acids:  82\n",
      "\t Molecular Weight:  9260.769340000003 Da\n",
      "\n",
      " Name:   Night heron coronavirus HKU19\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MEDYVITIPAIYILVAVILLGCIVFLLLIKTCLACCDFFTRTYTCILHTV....\n",
      "\t Number of Amino acids:  82\n",
      "\t Molecular Weight:  9567.095909999998 Da\n",
      "\n",
      " Name:   Wigeon coronavirus HKU20\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MSSWDVVIPGSVVIAFLIVACCILLLLFIKTCLAIFDCCSRSCTCVLDTV....\n",
      "\t Number of Amino acids:  86\n",
      "\t Molecular Weight:  9590.869119999998 Da\n",
      "\n",
      " Name:   Common moorhen coronavirus HKU21\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MVDNWDITIPGDYVIAALVVICVAVFLLFINTCLACIKLVYKCYKGATVL....\n",
      "\t Number of Amino acids:  82\n",
      "\t Molecular Weight:  9180.782620000002 Da\n",
      "\n",
      " Name:   Bat coronavirus BM48-31/BGR/2008\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MYSFVSEETGTLIVNSVLLFLAFVVFLLVTLAILTALRLCAYCCNIVNVS....\n",
      "\t Number of Amino acids:  76\n",
      "\t Molecular Weight:  8477.528989999999 Da\n",
      "\n",
      " Name:   Turkey coronavirus\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MTNLLSKSLEENGSFLTAVYIFVAFVALYLLGRVLQAFVQTADACCLFWY....\n",
      "\t Number of Amino acids:  99\n",
      "\t Molecular Weight:  11084.708779999994 Da\n",
      "\n",
      " Name:   Bat coronavirus 1A\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MLTLVDDHGLVINAILWLLFCVSVIIICCALIQLVQLCFICHRLCSNTVY....\n",
      "\t Number of Amino acids:  74\n",
      "\t Molecular Weight:  8522.490669999997 Da\n",
      "\n",
      " Name:   Scotophilus bat coronavirus 512\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MLQLVNDNGVVVNAILWLFVLFFVLVISITFVQLINLCFTCHRLCNNVVY....\n",
      "\t Number of Amino acids:  76\n",
      "\t Molecular Weight:  8714.707919999995 Da\n",
      "\n",
      " Name:   Rousettus bat coronavirus HKU9\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MYDIVGTNNSILIANVLVLIIICLLVVIVGCALLLILQFVFGVCGFVFKF....\n",
      "\t Number of Amino acids:  79\n",
      "\t Molecular Weight:  8856.820529999997 Da\n",
      "\n",
      " Name:   Human coronavirus NL63\n",
      "\n",
      " \t Protein Analysis: \n",
      "--------------------------\n",
      "\n",
      " \t Amino Acid Chain:  MFLRLIDDNGIVLNSILWLLVMIFFFVLAMTFIKLIQLCFTCHYFFSRTL....\n",
      "\t Number of Amino acids:  77\n",
      "\t Molecular Weight:  9153.824909999996 Da\n"
     ]
    }
   ],
   "source": [
    "fasta_records_basic_ds = file_into_fasta_records(f)\n",
    "print(\"\\nBasic Dataset \\n\\n\")\n",
    "fasta_records_analysis(fasta_records_basic_ds)\n",
    "\n",
    "fasta_records_related_ds = file_into_fasta_records(f2)\n",
    "print(\"\\nRelated Dataset \\n\\n\")\n",
    "fasta_records_analysis(fasta_records_related_ds)\n",
    "                \n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Split the datasets into training and test splits. I manually had stratify the data to make sure it was a valid set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File already exists, will continue\n",
      "File already exists, will continue\n"
     ]
    }
   ],
   "source": [
    "eighty_percent_basic_ds = open('eighty_percent_basic_ds.fasta', 'w')\n",
    "twenty_percent_basic_ds = open('twenty_percent_basic_ds.fasta', 'w')\n",
    "\n",
    "try:\n",
    "    os.mkdir(\"split_datasets_basic\")\n",
    "except FileExistsError: \n",
    "    print(\"File already exists, will continue\")\n",
    "    \n",
    "os.rename('eighty_percent_basic_ds.fasta', \"split_datasets_basic/eighty_percent_basic_ds.fasta\")\n",
    "os.rename('twenty_percent_basic_ds.fasta', \"split_datasets_basic/twenty_percent_basic_ds.fasta\")\n",
    "\n",
    "eighty_percent_length = int(len(fasta_records_basic_ds) * .8)\n",
    "twenty_percent_length = len(fasta_records_basic_ds) - eighty_percent_length\n",
    "\n",
    "write_fasta(fasta_records_basic_ds, eighty_percent_length, eighty_percent_basic_ds, True)\n",
    "write_fasta(fasta_records_basic_ds, len(fasta_records_basic_ds) - twenty_percent_length, twenty_percent_basic_ds, False)\n",
    "\n",
    "\n",
    "eighty_percent_basic_ds.close()\n",
    "twenty_percent_basic_ds.close()\n",
    "\n",
    "eighty_percent_related_ds = open('eighty_percent_related_ds.fasta', 'w')\n",
    "twenty_percent_related_ds = open('twenty_percent_related_ds.fasta', 'w')\n",
    "\n",
    "try:\n",
    "    os.mkdir(\"split_datasets_related\")\n",
    "except FileExistsError: \n",
    "    print(\"File already exists, will continue\")\n",
    "    \n",
    "os.rename('eighty_percent_related_ds.fasta', \"split_datasets_related/eighty_percent_basic_ds.fasta\")\n",
    "os.rename('twenty_percent_related_ds.fasta', \"split_datasets_related/twenty_percent_basic_ds.fasta\")\n",
    "\n",
    "eighty_percent_length = int(len(fasta_records_related_ds) * .8)\n",
    "twenty_percent_length = len(fasta_records_related_ds) - eighty_percent_length\n",
    "\n",
    "\n",
    "write_fasta(fasta_records_related_ds, eighty_percent_length, eighty_percent_related_ds, True)\n",
    "write_fasta(fasta_records_related_ds, len(fasta_records_related_ds) - twenty_percent_length, twenty_percent_related_ds, False)\n",
    "\n",
    "\n",
    "eighty_percent_related_ds.close()\n",
    "twenty_percent_related_ds.close()\n",
    "\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
